tensorflow_probability/python/experimental/mcmc/particle_filter.py

```diff
@@ -354,130 +354,125 @@ def particle_filter(observations,
-        observations=observations,
-        observation_fn=observation_fn,
-        initial_state_prior=initial_state_prior,
-        initial_state_proposal=initial_state_proposal,
-        num_particles=num_particles,
-        seed=init_seed)
-    propose_and_update_log_weights_fn = (
-        _particle_filter_propose_and_update_log_weights_fn(
-            observations=observations,
-            transition_fn=transition_fn,
-            proposal_fn=proposal_fn,
-            observation_fn=observation_fn,
-            num_transitions_per_observation=num_transitions_per_observation))
-
-    kernel = smc_kernel.SequentialMonteCarlo(
+    if initial_state_proposal is None:
+        initial_state_proposal = initial_state_prior
+
+    # Initializing particles and weights
+    initial_particles, initial_log_weights = \
+        initial_state_proposal.experimental_sample_and_log_prob(
+            [num_particles], seed=seed
+        )
+    initial_log_weights = (
+            initial_state_prior.log_prob(initial_particles) -
+            initial_state_proposal.log_prob(initial_particles)
+    )
+    initial_log_weights = tf.nn.log_softmax(initial_log_weights, axis=0)
+
+    if proposal_fn is None:
+        proposal_fn = transition_fn
+
+    def propose_and_update_log_weights_fn(step, state, seed):
+        """Particle filter propose and update for single steps"""
+        particles, log_weights = (
+            proposal_fn(step, state.particles).experimental_sample_and_log_prob(seed=seed)
+        )
+
+        assertions = _assert_batch_shape_matches_weights(
+            distribution=transition_fn(step, particles),
+            weights_shape=ps.shape(log_weights),
+            diststr='transition'
+        )
+        if proposal_fn(step, particles) != transition_fn(step, particles):
+            assertions += _assert_batch_shape_matches_weights(
+                distribution=transition_fn(step, particles),
+                weights_shape=ps.shape(log_weights),
+                diststr='proposal'
+            )
+
+        log_weights = (
+                observation_fn(step, state.particles).log_prob(tf.gather(observations, step))
+                + transition_fn(step, state.particles).log_prob(particles)
+                - log_weights)
+
+        log_weights = tf.nn.log_softmax(log_weights, axis=0)
+        with tf.control_dependencies(assertions):
+            return smc_kernel.WeightedParticles(
+                particles=particles,
+                log_weights=log_weights + _compute_observation_log_weights(
+                    step + 1, particles, observations, observation_fn,
+                    num_transitions_per_observation=num_transitions_per_observation)
+            )
+
+
+    def sequential_monte_carlo(
+            initial_state,
+            propose_and_update_log_weights_fn,
+            condition_fn,
+            resample_fn,
+            resample_criterion_fn,
+            trace_fn
+    ):
```

Comment:
Do we also need:
 * trace_criterion_fn:  ... same as for `particle_filter` ...
 * static_trace_allocation_size:  ... same as for `particle_filter` ...
 * parallel_iterations:  ... same as for `particle_filter` ...
 * seed:  ... same as for `particle_filter` ...
 * name:  ... same as for `particle_filter` ...